### DISCUSSION TITLE
on_page_markdown does not include snippet content

### DISCUSSION THREAD

**Discussion by vermiculus** [facelessuser/pymdown-extensions#2074] [Q&A] [answered]
### Description

MkDocs' hook for `on_page_markdown` does not include snippet content ([reference docs](https://www.mkdocs.org/dev-guide/plugins/#on_page_markdown)). The `markdown` value provided to the hook contains the snippet syntax (`--8<--`) and not the included content.

I'm not entirely sure whether this is an issue downstream (here) or upstream (mkdocs), but I'm starting here since you're going to be more familiar with where pymdown-extensions fits into MkDocs and whether you'd expect the current implementation to be subject to the hooks defined by MkDocs.

### Minimal Reproduction

1. Have a README that includes another file
2. Include an `on_page_markdown` hook that would alter the markdown content of the included file
3. Run `mkdocs build`

```markdown
# docs/README.md
the following content is included

--8<-- "docs/included.md"
```

```markdown
# docs/included.md
hello, world
```

```yaml
# mkdocs.yml
...
hooks:
  - my_hooks.py
...
```

```python
# my_hooks.py
def on_page_markdown(markdown, page, config, files):
    return markdown.replace("hello", "goodbye")
```

### Version(s) & System Info

- Operating System: macOS 13.2.1
- Python Version: 3.11.3
- Package Version: MkDocs 1.4.2, pymdown-extensions 9.9.2

**Comment by facelessuser** [comment, marked as answer]
I'm not exactly sure what the complaint is here. Per the MkDocs [documentation it states](https://www.mkdocs.org/dev-guide/plugins/#on_page_markdown):

> The `page_markdown` event is called after the page's markdown is loaded from file and can be used to alter the Markdown source text. The meta-data has been stripped off and is available as `page.meta` at this point.

So, it loads the Markdown source, which contains `--8<--` and allows you to alter it.  What is happening is that the the `on_page_markdown` event  in MkDocs let's you alter the Markdown source _before_ the Python Markdown parser processes the content, and the Snippets extension is a Python Markdown extension, not a MkDocs plugin. Hopefully that makes sense.

**Reply by vermiculus** [reply]
The problem is that the hook is never sent the content of the included file in a way that allows me to alter its content. So, short of re-implementing the snippets extension within my hook, I'm a bit SOL at the moment.

I think I'm understanding more about the content pipeline here (with your explanation, thanks 👍) – so it would seem that either
1. mkdocs doesn't (yet) have a hook that runs before markdown is rendered but *after* preprocessors (like snippets) run
2. mkdocs is invoking `on_page_markdown` in the 'wrong' place, which can't change at this point without breaking *someone's* site build

Is that a reasonable understanding? If so, I can move (1) to an upstream issue.

Thanks for your time 😄

**Reply by facelessuser** [reply]
Keep in mind that Snippets is a preprocessor, yes, but a preprocessor extension in Python Markdown. Python Markdown has no knowledge of what MkDocs is doing. I have no idea what you are trying to do either in regards to MkDocs and Snippets as that context has not been provided here.

But yes, if you absolutely needed the Snippets content in `on_page_markdown`, you'd almost have to implement your own approach, or process the content in post, but depending on what you are doing, post may be too late.

**Reply by facelessuser** [reply]
To answer your bullet points more directly:

1. The hook runs before Python Markdown does its work. Save MkDocs injecting its own extension in the Python Markdown pipeline ensuring it is the last preprocessor extension to be run, it will have no way of doing what you want. I don't know if this is anything they'd want to do or not.
2. Incorrect, it is an event run directly before Markdown is processed and gives you the content that will be passed to Python Markdown. That is its intention and exactly what it is doing, it just may not be what you were specifically hoping for.

**Reply by vermiculus** [reply]
Yes, I'm aware of the relationship between python-markdown and mkdocs (in that mkdocs is a consumer of python-markdown and python-markdown (and its extensions) have no knowledge of mkdocs).

(1) and (2) were either/or options – it was assumed that only one could be true :-)

I'll raise your potential approach for (1) to mkdocs. It seems this would be the only realistic way to get at the markdown source after pre-processing – since I assume python-markdown doesn't have a 'run pre-processors only' runmode. But I'm hardly the expert here :-)

Thanks for all your help! I'm not sure how to close discussions (they've been added to github since I was active here), but consider this resolved as far as pymdown-extensions is concerned.

**Reply by facelessuser** [reply]
It's probably still not going to work as you hope. One thing that should be noted is that fenced code blocks are run as preprocessor. This is due to some awkward limitations of how the Python Markdown parser works. That is an example of something that is a preprocessor that may not expect to be a preprocessor, and may be surprising in `on_page_markdown`.

**Reply by vermiculus** [reply]
Likely not, but it will make my actual use case (implementing basic shortlinks – a la [this example from git-commit(1)][1]) possible for included files.

[1]: https://github.com/git/git/blob/fe86abd7511a9a6862d5706c6fa1d9b57a63ba09/Documentation/git-commit.txt#L27

**Comment by facelessuser** [comment]
Moving to discussions
